data_structures/monotonic_stack.cpp

```diff
@@ -0,0 +1,107 @@
+/**
+ * @file
+ * @brief Implementation of Monotonic Stack
+ * @details
+ * This is an implementation of a monotonic stack which maintains the stack in increasing order.
+ * Elements are pushed such that the stack remains monotonic.
+ * If an element is pushed that is smaller than the top element, the top elements are popped until the correct position is found.
+ * @see https://en.wikipedia.org/wiki/Monotonic_Stack
+ * @author [Srijan Baniyal](https://github.com/Srijan-Baniyal)
+ */
+
+#include <iostream>        /// For IO operations
+#include <vector>         /// For std::vector
+#include <stdexcept>      /// For std::out_of_range
+
+/**
+ * @namespace data_structures
+ * @brief A namespace for various data structure implementations
+ */
+namespace data_structures {
+
+/**
+ * @class MonotonicStack
+ * @brief A class representing a monotonic stack which maintains elements in increasing order.
+ */
+class MonotonicStack {
+ private:
+    std::vector<int> stack;   ///< Vector to store stack elements
```

Comment:
could this not be generic to store values other than int? since all we are doing is comparing if input > last_val?

Reply:
instead of vector<int> stack . I can do std::vector<T> stack;

will this work out for you ??

Reply:
yes the class should be Templated.

Reply:
👌👌👌